### PR TITLE
Consolidate Ingress resource

### PR DESCRIPTION
**Pull request by fsat**
The Nginx and Istio ingress resource are quite similar - as such we'll generalize the ingress resource generation and provide the following customisation points:

* [x] Allow user to create Ingress annotations which is a `String` key/value pair.
* [x] Allow user to specify path append expression - this is support wildcard style path matching ala Istio (not done)

### CODE REVIEW

**Comment by fsat** [comment]
Marked as wip - pending manual test.

**Comment by fsat** [comment]
Manual test is completed successfully.

**Default args**

With default argument the following ingress option is generated (last line):

```
192-168-1-5:reactive-cli felixsatyaputra$ ./cli/target/scala-2.11/rp generate-deployment fsat/testimpl:1.0.0-SNAPSHOT --target kubernetes --kubernetes-version 1.6
---
{"apiVersion":"apps/v1beta1","kind":"Deployment","metadata":{"name":"testImpl-v1.0.0-SNAPSHOT","labels":{"app":"testImpl","appVersionMajor":"testImpl-v1","appVersionMajorMinor":"testImpl-v1.0","appVersion":"testImpl-v1.0.0-SNAPSHOT"}},"spec":{"replicas":1,"serviceName":"testImpl-v1","template":{"app":"testImpl","appVersionMajor":"testImpl-v1","appVersionMajorMinor":"testImpl-v1.0","appVersion":"testImpl-v1.0.0-SNAPSHOT"},"spec":{"containers":[{"name":"testImpl","image":"fsat/testimpl:1.0.0-SNAPSHOT","imagePullPolicy":"IfNotPresent","env":[{"name":"FOO","value":"123"}],"ports":[{"containerPort":16668,"name":"ta-v1"}],"readinessProbe":{"httpGet":{"path":"/healthz","port":1234},"periodSeconds":60},"livenessProbe":{"exec":{"command":["/bin/bash","-c","exit 0"]}}}]}}}
---
{"apiVersion":"v1","kind":"Service","metadata":{"labels":{"app":"testImpl"},"name":"testImpl"},"spec":{"clusterIP":"None","ports":[{"name":"ta-v1","port":16668,"protocol":"TCP","targetPort":16668}],"selector":{"app":"testImpl"}}}
---
{"apiVersion":"extensions/v1beta1","kind":"Ingress","metadata":{"name":"testImpl"},"spec":{"rules":[{"http":{"paths":[{"path":"^/ta.*$","backend":{"serviceName":"ta-v1","servicePort":16668}}]}}]}}
```

The formatted ingress JSON looks like this:

```
{
	"apiVersion": "extensions/v1beta1",
	"kind": "Ingress",
	"metadata": {
		"name": "testImpl"
	},
	"spec": {
		"rules": [{
			"http": {
				"paths": [{
					"path": "^/ta.*$",
					"backend": {
						"serviceName": "ta-v1",
						"servicePort": 16668
					}
				}]
			}
		}]
	}
}
```

**Ingress specific args**

With ingress specific args `--kubernetes-ingress-annotation foo=bla --kubernetes-ingress-path-append XYZ`, the following ingress resource is generated (last line):

```
192-168-1-5:reactive-cli felixsatyaputra$ ./cli/target/scala-2.11/rp generate-deployment fsat/testimpl:1.0.0-SNAPSHOT --target kubernetes --kubernetes-version 1.6 --kubernetes-ingress-annotation foo=bla --kubernetes-ingress-path-append XYZ
---
{"apiVersion":"apps/v1beta1","kind":"Deployment","metadata":{"name":"testImpl-v1.0.0-SNAPSHOT","labels":{"app":"testImpl","appVersionMajor":"testImpl-v1","appVersionMajorMinor":"testImpl-v1.0","appVersion":"testImpl-v1.0.0-SNAPSHOT"}},"spec":{"replicas":1,"serviceName":"testImpl-v1","template":{"app":"testImpl","appVersionMajor":"testImpl-v1","appVersionMajorMinor":"testImpl-v1.0","appVersion":"testImpl-v1.0.0-SNAPSHOT"},"spec":{"containers":[{"name":"testImpl","image":"fsat/testimpl:1.0.0-SNAPSHOT","imagePullPolicy":"IfNotPresent","env":[{"name":"FOO","value":"123"}],"ports":[{"containerPort":16668,"name":"ta-v1"}],"readinessProbe":{"httpGet":{"path":"/healthz","port":1234},"periodSeconds":60},"livenessProbe":{"exec":{"command":["/bin/bash","-c","exit 0"]}}}]}}}
---
{"apiVersion":"v1","kind":"Service","metadata":{"labels":{"app":"testImpl"},"name":"testImpl"},"spec":{"clusterIP":"None","ports":[{"name":"ta-v1","port":16668,"protocol":"TCP","targetPort":16668}],"selector":{"app":"testImpl"}}}
---
{"apiVersion":"extensions/v1beta1","kind":"Ingress","metadata":{"name":"testImpl","annotations":{"foo":"bla"}},"spec":{"rules":[{"http":{"paths":[{"path":"^/ta.*$XYZ","backend":{"serviceName":"ta-v1","servicePort":16668}}]}}]}}
```

The formatted Ingress JSON looks like this:

```
{
	"apiVersion": "extensions/v1beta1",
	"kind": "Ingress",
	"metadata": {
		"name": "testImpl",
		"annotations": {
			"foo": "bla"
		}
	},
	"spec": {
		"rules": [{
			"http": {
				"paths": [{
					"path": "^/ta.*$XYZ",
					"backend": {
						"serviceName": "ta-v1",
						"servicePort": 16668
					}
				}]
			}
		}]
	}
}
```

File: cli/src/test/scala/com/lightbend/rp/reactivecli/runtime/kubernetes/KubernetesPackageTest.scala [hunk 1]
```diff
@@ -302,11 +227,11 @@ object KubernetesPackageTest extends TestSuite {
                 |	}
                 |}
               """.stripMargin.parse.right.get
-
+            println(ingress.payload)
```
Comment (longshorej) [inline cli/src/test/scala/com/lightbend/rp/reactivecli/runtime/kubernetes/KubernetesPackageTest.scala]: I assume this was committed by accident so I'll remove it.

Comment (fsat) [reply cli/src/test/scala/com/lightbend/rp/reactivecli/runtime/kubernetes/KubernetesPackageTest.scala]: Thanks Jason 😳

**Pull request merged**